File: src/main/scala/com/vc/easy/L543.scala
package com.vc.easy

class TreeNode543(var _value: Int) {
  var value: Int = _value
  var left: TreeNode543 = null
  var right: TreeNode543 = null
}

object L543 {
  def diameterOfBinaryTree(root: TreeNode543): Int = {
    var diameter = 0
    def height(root:TreeNode543): Int = {
      if(root != null) {
        val l = height(root.left)
        val r = height(root.right)
        diameter = Math.max(diameter, l + r)
        return 1 + Math.max(l, r)
      }
      0
    }
    height(root)
    diameter
  }
}

File: src/main/java/com/vc/easy/L543.java
package com.vc.easy;

import com.vc.hard.TreeNode;

class L543 {
    int diameter = 0;
    public int diameterOfBinaryTree(TreeNode root) {
        height(root);
        return diameter;
    }

    private int height(TreeNode root) {
        if(root == null) return 0;
        int left = height(root.left);
        int right = height(root.right);
        int height = 1 + Math.max(left, right);
        diameter = Math.max(diameter, left + right);
        return height;
    }
}
